### ISSUE TITLE
dashboard sends traffic to api-server directly rather than to the load balancer in a multi-master setup

### ISSUE THREAD

**Issue by captainblerko** [kubernetes/dashboard#830]
#### Issue details

In a cluster with multi-masters running api-server behind a load balancer, the dashboard connects directly to master servers. I do have issues with the dashboard container resolving DNS properly in my cluster, but I don't think this is the sole problem (and it would be nice to be able to solve that in the same way as I can with the kubelet container under coreos: https://coreos.com/kubernetes/docs/latest/kubelet-wrapper.html see customizing rkt options). I created a sidecar container containing kubectl running in proxy mode and that talks to the load balancer for the api-servers and paired it with dashboard. I set the --apiserver-host to be localhost:8080 and that works, but I still see traffic going from the node hosting the dashboard to the master servers directly rather than via the kubectl proxy/load balancer.
##### Environment

Kubernetes CoreOS cluster  running in coreos

```
Dashboard version: v1.1.0-beta1
Kubernetes version: 1.2
Operating system: coreOS 899.17.0
```
##### Steps to reproduce
1. Have a working multi-master cluster with api-servers on those masters fronted by a load balancer. In our case this is an ELB on AWS.
2. Modify the dashboard yaml file to include a sidecar kubectl container running in proxy mode on port 8080
3. Point the --apiserver-host in the dashboard configuration to the localhost:8080
##### Observed result

Running tcpdump on the node running the dashboard you still see connections to the master ip addresses on port 443 
##### Expected result

I expected all api-server traffic to be routed via the load balancer

**Comment by bryk** [comment]
Hmm... This should work. Can you clone the repository and run `./build/run-gulp-in-docker.sh docker-image:canary`. Then you'll have a docker image of latest HEAD version. Can you run it to determine whether problem still persists?

**Comment by bryk** [comment]
Btw, how does your YAML look like? Can you show it?

**Comment by captainblerko** [comment]
I will have a go at cloning the repo in a bit, I threw away my yaml file so I will just try to recreate it, bear with me.

**Comment by captainblerko** [comment]
BTW I believe the dashboard does talk via the kubectl proxy mostly, just not exclusively. The problem I am rally trying to solve is locking down the connections between cluster elements.

**Comment by captainblerko** [comment]
Here is the yaml, the only bits I added are:
```
        - image: <localregistry>kubectl:latest
          name: kubectl-proxy
          imagePullPolicy: Always
```
and

```
        imagePullSecrets:
          - name: registry-pull-secret
```

Modified this line to be:
```
            - --apiserver-host=http://localhost:8080
```

[kubernetes-dashboard-sidecar.txt](https://github.com/kubernetes/dashboard/files/300163/kubernetes-dashboard-sidecar.txt)

**Comment by captainblerko** [comment]
Just checked that I see the same behaviour with that new version of the sidecar yaml and I do, every 30 seconds the dashboard checks in with one of my masters. If I kill the pod the traffic stops.

**Comment by captainblerko** [comment]
I built the image but it came out as 1.5GB in size that doesn't seem right, can you point me in the direction of where I went wrong?

**Comment by bryk** [comment]
The development docker image is 1.5 Gigs, but there should be other image produced. Named, like, `kubernetes-dashboard-amd64:canary`. Do you have it in `docker images`?

**Comment by captainblerko** [comment]
Ah ok, no docker images just lists the biggie :-) Let me try again

**Comment by bryk** [comment]
Any success?

**Comment by captainblerko** [comment]
Sorry got busy, just trying the rebuild now

**Comment by captainblerko** [comment]
I get:
Task '.' is not in your gulpfile
There is a space between canary and the . in you command above isn't there?

**Comment by bryk** [comment]
The command is:
```bash
./build/run-gulp-in-docker.sh docker-image:canary
```

But anyways. Tomorrow we'll release next beta. It has logging enabled. You can use it for tests. I'll let you know.

**Comment by captainblerko** [comment]
OK getting rid of the spurious period got the canary built, I deployed it and it behaves in the same way talking to the master server every 30s

**Comment by bryk** [comment]
How do you know that it talks to the master every 30s? Do you have logs? Can you show logs of dashboard UI?

**Comment by captainblerko** [comment]
I am running on coreos and this provides a toolbox command which creates a container from which you can run diagnostic commands, I see the traffic by using tcpdump, I believe it is the dashboard container that creates the traffic as I see the connection reset when I tear down the container and no further traffic thereafter.

**Comment by bryk** [comment]
Can you run `docker logs <...>`? For dashboard container?

**Comment by captainblerko** [comment]
```
Starting HTTP server on port 9090
Creating API server client for https://<apiserver_ip>:443
Successful initial request to the apiserver, version: v1.2.0+coreos.1
Creating in-cluster Heapster client
Incoming HTTP/1.1 GET /api/v1/namespace request from <docker_ip>:57822
Getting namespace list
Outcoming response to <docker_ip>:57822 with 201 status code
Incoming HTTP/1.1 GET /api/v1/workload/int-ngineered request from <docker_ip>:57822
Getting lists of all workloads
Getting pod metrics
Outcoming response to <docker_ip>:57822 with 201 status code
Incoming HTTP/1.1 GET /api/v1/namespace request from <docker_ip>:57822
Getting namespace list
Outcoming response to <docker_ip>:57822 with 201 status code
Incoming HTTP/1.1 GET /api/v1/workload/kube-system request from <docker_ip>:57822
Getting lists of all workloads
Getting pod metrics
Outcoming response to <docker_ip>:57822 with 201 status code
Incoming HTTP/1.1 GET /api/v1/event/kube-system/kubernetes-dashboard request from <docker_ip>:57822
Getting events related to kubernetes-dashboard replication controller in kube-system namespace
Incoming HTTP/1.1 GET /api/v1/replicationcontroller/kube-system/kubernetes-dashboard request from <docker_ip>:39590
Getting details of kubernetes-dashboard replication controller in kube-system namespace
Getting pod metrics
Found 13 events related to kubernetes-dashboard replication controller in kube-system namespace
Outcoming response to <docker_ip>:57822 with 201 status code
Outcoming response to <docker_ip>:39590 with 201 status code
```
"apiserver_ip" is the default kubernetes api-server service ip, "docker_ip" is the ip of the docker0 interface on the node that the dashboard is running on

**Comment by floreks** [comment]
Closing as stale. Please reopen if there will be some issues with new versions of Dashboard. By default when using in-cluster config Dashboard talks to api-server via default kubernetes service that should route the trafic to one of the masters in HA config.
